src/box/txn_event_trigger.c

```diff
@@ -0,0 +1,340 @@
+{
+	txn_event->mode = TXN_EVENT_NO_TRIGGERS;
+	txn_event->space_id = 0;
+	txn_event->space_event = NULL;
+}
+
+void
+txn_event_add_space(struct txn *txn, struct space *space, int event_id)
+{
+	assert(space != NULL);
+	assert(event_id >= 0 && event_id < txn_event_id_MAX);
+	struct space_event *space_event = &space->txn_events[event_id];
+	if (likely(!space_event_has_triggers(space_event)))
+		return;
+
+	struct txn_event *txn_event = &txn->txn_events[event_id];
+	switch (txn_event->mode) {
+	case TXN_EVENT_NO_TRIGGERS:
+		txn_event->mode = TXN_EVENT_ONE_SPACE;
+		txn_event->space_id = space_id(space);
+		txn_event->space_event = space_event;
+		break;
+	case TXN_EVENT_ONE_SPACE:
+		if (txn_event->space_id != space_id(space)) {
+			txn_event->mode = TXN_EVENT_MULTIPLE_SPACES;
+			txn_event->space_id = 0;
+			txn_event->space_event = NULL;
+		}
+		break;
+	case TXN_EVENT_MULTIPLE_SPACES:
+		break;
+	default:
+		unreachable();
+	}
+}
+
+enum { SPACE_ID_FILTER_ALL_SPACES = -1 };
+
+/**
+ * State of the iterator, which is passed to txn_iterator_next().
+ */
+struct txn_iterator_state {
+	/** Request number, starting from 1. */
+	uint32_t req_num;
+	/** Current statement of the transaction. */
+	struct txn_stmt *stmt;
+	/**
+	 * Iterate only over statements with the given space id.
+	 * Or set to SPACE_ID_FILTER_ALL_SPACES for all spaces.
+	 */
+	int64_t space_id_filter;
+};
+
+/**
+ * The iterator goes through every statement of the transaction.
+ */
+static int
+txn_iterator_next(struct func_adapter *func, struct func_adapter_ctx *ctx,
+		  void *state)
+{
+	struct txn_iterator_state *txn_state =
+		(struct txn_iterator_state *)state;
+	struct txn_stmt *stmt = txn_state->stmt;
+
+	/* Skip unwanted spaces if space id filter is set. */
+	if (txn_state->space_id_filter != SPACE_ID_FILTER_ALL_SPACES) {
+		while (stmt != NULL &&
+		       space_id(stmt->space) != txn_state->space_id_filter) {
+			stmt = stailq_next_entry(stmt, next);
+		}
+	}
+
+	if (stmt == NULL)
+		return 0;
+	/*
+	 * The iterator returns 4 values:
+	 * 1. An ordinal request number;
+	 * 2. The old value of the tuple;
+	 * 3. The new value of the tuple;
+	 * 4. The ID of the space.
+	 */
+	func_adapter_push_double(func, ctx, txn_state->req_num++);
+	if (stmt->old_tuple != NULL)
+		func_adapter_push_tuple(func, ctx, stmt->old_tuple);
+	else
+		func_adapter_push_null(func, ctx);
+	if (stmt->new_tuple != NULL)
+		func_adapter_push_tuple(func, ctx, stmt->new_tuple);
+	else
+		func_adapter_push_null(func, ctx);
+	func_adapter_push_double(func, ctx, space_id(stmt->space));
+
+	txn_state->stmt = stailq_next_entry(stmt, next);
+	return 0;
+}
+
+/**
+ * Run triggers registered in `txn' for the `event'.
+ * `stmt' points to the first statement of a transaction or a savepoint.
+ * `can_abort' is set to true if the trigger can abort the transaction.
+ * `space_id' is passed to the iterator to filter spaces by the ID.
+ */
+static int
+run_triggers_for_event(struct txn *txn, struct txn_stmt *stmt,
```

Comment:
`run_triggers_for_event`
`run_triggers_for_txn_event`
`run_triggers_for_multi_space_event`
`run_triggers_for_space_event`

I still get confused sometimes by these names. All these functions are logically named, but there's a lot to keep in your head while reading this code - `run_triggers_for_event` runs triggers from event which are always triggers without filtration (for example, `box.on_commit`). And `run_triggers_for_txn_event` calls triggers from `txn_event`, which is actually not an event - it just shows how much spaces with triggers participate in transactions.

Pretty function names are not the primary goal, but it would be nice to try to reduce cognitive complexity of this submodule. For example, we could try to call the functions based on their purpose (`run_triggers_general`, `run_triggers_of_spaces` and so on).

I'll leave the final decision on that matter to the second reviewer and you.

Reply:
Renamed.